### ISSUE TITLE
ssl_ca_file option can not verify self signed certificate

### ISSUE THREAD

**Issue by ozenozkaya** [civetweb/civetweb#797]
In civetweb 1.11 version I implemented a rest server with TLS enabled and it works flawless except ssl_ca_file parameter. I want to accept only one trusted client certificate (of my frontend application). I created frontend certificates as it is described in /doc/OpenSSL.md document. But I added password protection as it is visible as below

openssl genrsa -aes256 -passout pass:aaa -out frontend.key 2048
openssl req -config openssl.cnf  -new -passin pass:aaa  -key frontend.key -out frontend.csr
cp frontend.key frontend.key.orig
openssl rsa -passin pass:aaa -in frontend.key.orig -out frontend.key
openssl x509 -passin pass:aaa frontend.csr -signkey frontend.key -out frontend.crt
cp frontend.crt frontend.pem
cat frontend.key >> frontend.pem

I create server certificate of civetweb in same way. In this case civetweb works fine but it can not verify frontend(client) certificate since it is self signed (and private key passphrase protected). Here below my civetweb options:

"ssl_certificate", "backend.pem",
"ssl_protocol_version", "4",
"ssl_verify_peer","yes",
"ssl_short_trust","yes",
"ssl_ca_file", "frontend.pem",
"access_control_list","+127.0.0.1", 
"enable_directory_listing","no",
"strict_transport_security_max_age", "15552000",
"ssl_cipher_list","AES256-SHA:AES128-SHA:AES128-GCM-SHA256",
"enable_auth_domain_check", "no"

In this configuration I see following entry in error.log file:
[1571752471] [error] [client 127.0.0.1] sslize error: error:1417C086:SSL routines:tls_process_client_certificate:certificate verify failed

Do you know how I can get this setup working? I also passed backend.pem to ssl_ca_file but it is not verifiable. I can verify frontend certificate with following openssl command:
openssl verify -CAfile frontend.pem frontend.pem 
frontend.pem: OK 

Why civetweb can not verify client certificate?

**Comment by bel2125** [comment]
I tested it using the certificates here: https://github.com/civetweb/civetweb/tree/master/resources/cert
they are all self signed.

There is a "make_cert" file, containing the command lines I uses to create them.

The passphrase is stripped from them - CivetWeb must be able to read the cert, and there is no point anyway in having a pass phrase if you need to provide it to the civetweb config in plain text - so this is the command to remove it:
https://github.com/civetweb/civetweb/blob/32299611c866536c55d55a8ee908c817d19fe769/resources/cert/make_certs.sh#L11

**Comment by bel2125** [comment]
> there is no point anyway in having a pass phrase if you need to provide it to the civetweb config in plain text 

well, one could query the pass phrase from the command line (Linux) or open some dialog (Windows) ... but this only works, if you run the server interactively, not as a service / not from autostart

Currently CivetWeb needs certs without a passphrase

**Comment by ozenozkaya** [comment]
Hello @bel2125, first of all thanks for all the clarification and hints. I totally understand the design decisions behind "paswordless private key" need. When I used the certificates you pointed I achieved mutual authentication flawless. 

But I want to note more information regarding to this issue for other who may need some advanced cases. If one creates a certificate with key usage, extended key usage and SAN, every parameter should be verifiable via OpenSSL. For example my certificates did not accepted by Civetweb since I used it for client authentication but in ExtendedKeyUsage field I did not put that information.

When I removed the passpharase as you highlighted and I changed extended key usage to below, I get everything worked.

basicConstraints=critical,CA:FALSE
keyUsage = nonRepudiation, digitalSignature, keyEncipherment, keyCertSign
extendedKeyUsage = serverAuth,clientAuth,codeSigning

This is the case for SSLv3 certificates.

So I will close the issue since it is not a bug but an intended system behavior. Again, sincerely thanks a lot for your support.
